### PR TITLE
Add new tool: Firecrawl

### PR DESCRIPTION
**Pull request by richards199999**
# Description

Since the combination of Bing Search and JinaReader is always slow, so I add this new tool that do search and scrap at the same time. [Official Intro](https://www.firecrawl.dev/)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

# TODO Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

File: api/core/tools/provider/builtin/firecrawl/tools/crawl.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+class CrawlTool(BuiltinTool):
+    def _invoke(self, user_id: str, tool_parameters: dict[str, Any]) -> Union[ToolInvokeMessage, list[ToolInvokeMessage]]:
+        url = "https://api.firecrawl.dev/v0/crawl"
+
+        # innitialize the app object with the api key
+        app = FirecrawlApp(api_key=self.runtime.credentials['firecrawl_api_key'])
+
+        # crawl the url
+        crawl_result = app.crawl_url(url, params=tool_parameters, wait_until_done=True, timeout=5)
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/firecrawl/tools/crawl.py]:
```suggestion
        crawl_result = app.crawl_url(url, params=tool_parameters, wait_until_done=True, timeout=60)

```

File: api/core/tools/provider/builtin/firecrawl/tools/crawl.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        
+        # reformat crawl result
+        crawl_output = "**Crawl Result**\n\n"
+        for result in crawl_result:
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/firecrawl/tools/crawl.py]: Need to handle the exception please use try catch

File: api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        human_description:
+          en_US: If enabled, the crawler will generate alt text for images using LLMs. This feature requires a paid plan.
+          zh_Hans: 如果启用，爬虫将使用LLMs为图像生成alt文本。此功能需要付费计划。
+        form: form
+        options:
+          - value: true
+            label:
+              en_US: Yes
+              zh_Hans: 是
+              pt_BR: Yes
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml]: All `pt_BR` can be removed for now.

**Review by crazywoola — CHANGES_REQUESTED** [review]

File: api/requirements.txt [hunk 4]
```diff
@@ -80,4 +80,4 @@ lxml==5.1.0
 xlrd~=2.0.1
 pydantic~=1.10.0
 pgvecto-rs==0.1.4
-oss2==2.15.0
```
Comment (takatost) [inline api/requirements.txt:83]: The oss2 package was mistakenly deleted here. Please restore it.

**Review by takatost — CHANGES_REQUESTED** [review]

**Comment by takatost** [comment]
@Yeuoly Can you help review this PR?

File: api/requirements.txt (hunk 4 above)
Comment (richards199999) [reply api/requirements.txt:83]: sure thing!

File: api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    type: string
+    required: false
+    label:
+      en_US: Crawler Options
+      zh_Hans: 爬虫选项
+    human_description:
+      en_US: Additional options to configure the web crawler behavior.
+      zh_Hans: 用于配置网页爬虫行为的附加选项。
+    form: form
+    children:
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml]: children is an invalid field, and I think we should only declare necessary parameters here. BTW, parameter list is a flat field, and nested structure is not allowed, I can help you with this PR if you don't mind some commits to be pushed into your PR.

**Review by Yeuoly — CHANGES_REQUESTED** [review]

File: api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml (hunk 5 above)
Comment (richards199999) [reply api/core/tools/provider/builtin/firecrawl/tools/crawl.yaml]: Sure thing, thank you!😭😭😭 @Yeuoly

**Review by Yeuoly — APPROVED** [review]
LGTM

**Comment by richards199999** [comment]
Thank you!!!!!!!😭😭😭😭 @Yeuoly

**Comment by BradKML** [comment]
@richards199999 Does this include the localized implementation rather than the external API implementation? Apparently Ollama can be used as a drop-in replacement https://github.com/mendableai/firecrawl/discussions/195

**Pull request merged**